src/safeposix/syscalls/fs_calls.rs

```diff
@@ -988,118 +1036,122 @@ impl Cage {
+                },
+            Pipe(pipe_filedesc_obj) => {
+                let pipenumber = pipe_filedesc_obj.pipe;
+                let read_references = Self::_lookup_refs_by_pipe_end(self, pipenumber, O_RDONLY);
+                let write_references = Self::_lookup_refs_by_pipe_end(self, pipenumber, O_WRONLY);
 
-                    if write_references == 1 && pipe_filedesc_obj.flags == O_WRONLY {
-                        // let pipetable.pipenumber.eof = true;
-                    }
+                if write_references == 1 && pipe_filedesc_obj.flags == O_WRONLY {
+                    // let pipetable.pipenumber.eof = true;
+                }
 
-                    if read_references + write_references == 1 {
-                        //del pipetable.pipenumber
-                    }
-                },
-                //TO DO: check IS_EPOLL_FD and if true, call epoll_object_deallocator (if necessary?)
-                File(normalfile_filedesc_obj) => {
-                    let inodenum = normalfile_filedesc_obj.inode;
-                    let inodeobj = mutmetadata.inodetable.get_mut(&inodenum).unwrap();
-                    let mut fobjtable = FILEOBJECTTABLE.write().unwrap();
+                if read_references + write_references == 1 {
+                    //del pipetable.pipenumber
+                }
+            },
+            //TO DO: check IS_EPOLL_FD and if true, call epoll_object_deallocator
+            File(normalfile_filedesc_obj) => {
+                let inodenum = normalfile_filedesc_obj.inode;
+                let inodeobj = mutmetadata.inodetable.get_mut(&inodenum).unwrap();
+                let mut fobjtable = FILEOBJECTTABLE.write().unwrap();
 
-                    match inodeobj {
-                        Inode::File(ref mut normalfile_inode_obj) => {
-                            normalfile_inode_obj.refcount -= 1;
+                match inodeobj {
+                    Inode::File(ref mut normalfile_inode_obj) => {
+                        normalfile_inode_obj.refcount -= 1;
 
-                            //if there are still references to the file, then do nothing to the file
-                            if normalfile_inode_obj.refcount != 0 {
-                                return 0;
-                            }
-                            if fobjtable.contains_key(&inodenum) {
-                                //close and delete the file in the object table
-                                fobjtable.get(&inodenum).unwrap().close().unwrap();
-                                fobjtable.remove(&inodenum);
+                        //if it's not a reg file, then we have nothing to close
+                        if !is_reg(normalfile_inode_obj.mode) {
+                            match fobjtable.get(&inodenum) {
+                                Some(_) => {syscall_error(Errno::ENOEXEC, "close or dup", "Non-regular file in file object table");},
+                                None => {return 0;}
```

Comment:
You still want to do the rest of the close stuff for non-regular files--like removing from fdtable and removing the inode if the refcount and linkcount are 0

Reply:
And anyway this whole block is unnecessary because all files that have inodes that are Inode::File are assumed to be reg